### PR TITLE
[Fun-ASR] Added bucketed CUDA graphs for encoder forward

### PR DESCRIPTION
**Pull request by WilsonZheng0327**
## Motivation

Fun-ASR is very encoder-heavy and launch-bound. Implementing torch.compile() gave little improvements which was surprising, hence I did some profiling (images are from concurrency-32).

image

Here `prelm.forward` is the encoder forwarding, which is extremely launch-bound. Syncing takes ~20 microseconds while the entire forward takes 150ms.

Hence, I wanted to add CUDA graphs to see if that would help. Since the batch size and T size varies a lot in audio models, bucketing based on batch and T with padding makes sense.

**After PR**

image

After warmup, encoder takes 7ms (compared to 150ms+).

## Modifications

- Add `FunASREncoderCudaGraphRunner` (`sglang_omni/models/fun_asr/encoder_cuda_graph.py`): capture-once/replay per (batch, LFR-length) bucket — batch padded up to {1, 2, 4, 8} (capped at `pre_lm_max_batch_size`), LFR frames padded up to multiples of 64 (max 512 ≈ the 30 s clip limit), ≤ ~32 graphs total, captured lazily on first use with a shared memory pool and VRAM headroom check.
- Compute the SANM pad mask **inside** the capture from a static lengths tensor, so each replay only copies input features and lengths; padded batch rows use `ilens=1` (one valid zero frame) so fully-masked SDPA rows cannot produce NaN, and their output rows are discarded.
- Use `capture_error_mode="thread_local"` so the LM scheduler thread's concurrent kernel launches cannot poison a capture; serialize capture/replay behind a lock (both the pre-LM worker and inline prefill can reach `get_audio_feature`).
- Route `get_audio_feature` through the runner when attached, with unchanged eager fallback on any capture failure, oversized shape, or low VRAM (per-bucket failure tracking; the feature degrades to current behavior, never errors).
- Wire an `enable_encoder_cuda_graph` flag (default **off**) through `create_sglang_fun_asr_executor` → `FunASREngineBuilder.setup_model_resources`; it supersedes `enable_encoder_torch_compile` with a logged warning (a dynamo-compiled callable cannot be captured).
- Unit tests: bucket rounding, runner routing, eager fallback, and unchanged no-runner behavior (`tests/unit_test/fun_asr/test_encoder_cuda_graph.py`).

## Benchmarks

Run on H200, seedtts 1088 samples

### Concurrency-32 results (no embedding cache)

| Metric | Baseline (eager) | torch.compile | CUDA graphs (this PR) |
|---|---:|---:|---:|
| Throughput | 56.31 samples/s | 72.22 samples/s | 130.12 samples/s |
| Improvement vs baseline | — | **+28.3% / 1.283×** | **+131.1% / 2.311×** |
| Wall time | 19.33 s | 15.07 s | 8.36 s |
| Mean latency | 561.4 ms | 440.6 ms | 244.7 ms |
| p95 latency | 659.4 ms | 626.7 ms | 330.2 ms |
| p99 latency | 695.9 ms | 755.3 ms | 394.6 ms |
| Mean RTF | 0.12338 | 0.09601 | 0.05304 |
| Corpus WER | 0.01710 | 0.01740 | 0.01718 |
| Skipped | 0 | 0 | 0 |

### Full concurrency sweep (no embedding cache)

| Conc. | Variant | Throughput | Tput vs baseline | Wall time | Mean latency | p95 latency | p99 latency | Mean RTF | Corpus WER | Skipped |
|---:|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| 1 | baseline | 12.98/s | — | 83.84 s | 76.9 ms | 87.8 ms | 96.9 ms | 0.01675 | 0.01710 | 0 |
| 1 | torch.compile | 14.15/s | **+9.1% / 1.09×** | 76.87 s | 70.5 ms | 81.5 ms | 90.2 ms | 0.01535 | 0.01740 | 0 |
| 1 | CUDA graphs | 16.38/s | **+26.2% / 1.26×** | 66.42 s | 60.9 ms | 71.9 ms | 80.1 ms | 0.01323 | 0.01718 | 0 |
| 2 | baseline | 25.26/s | — | 43.07 s | 79.0 ms | 129.4 ms | 154.3 ms | 0.01722 | 0.01710 | 0 |
| 2 | torch.compile | 24.50/s | **-3.0% / 0.97×** | 44.41 s | 81.5 ms | 125.3 ms | 139.2 ms | 0.01778 | 0.01740 | 0 |
| 2 | CUDA graphs | 34.78/s | **+37.7% / 1.38×** | 31.28 s | 57.3 ms | 73.1 ms | 85.6 ms | 0.01248 | 0.01718 | 0 |
| 4 | baseline | 35.86/s | — | 30.35 s | 111.1 ms | 166.4 ms | 197.4 ms | 0.02430 | 0.01710 | 0 |
| 4 | torch.compile | 40.55/s | **+13.1% / 1.13×** | 26.83 s | 98.3 ms | 127.5 ms | 148.5 ms | 0.02147 | 0.01740 | 0 |
| 4 | CUDA graphs | 60.87/s | **+69.8% / 1.70×** | 17.88 s | 65.4 ms | 79.1 ms | 92.5 ms | 0.01426 | 0.01718 | 0 |
| 8 | baseline | 53.56/s | — | 20.32 s | 148.9 ms | 205.5 ms | 233.9 ms | 0.03255 | 0.01710 | 0 |
| 8 | torch.compile | 60.67/s | **+13.3% / 1.13×** | 17.93 s | 131.6 ms | 190.2 ms | 219.1 ms | 0.02881 | 0.01740 | 0 |
| 8 | CUDA graphs | 93.47/s | **+74.5% / 1.75×** | 11.64 s | 85.3 ms | 105.0 ms | 134.7 ms | 0.01864 | 0.01718 | 0 |
| 16 | baseline | 56.58/s | — | 19.23 s | 281.2 ms | 344.8 ms | 370.8 ms | 0.06169 | 0.01710 | 0 |
| 16 | torch.compile | 64.98/s | **+14.8% / 1.15×** | 16.75 s | 244.5 ms | 324.5 ms | 378.0 ms | 0.05345 | 0.01740 | 0 |
| 16 | CUDA graphs | 118.91/s | **+110.2% / 2.10×** | 9.15 s | 133.4 ms | 195.7 ms | 224.2 ms | 0.02897 | 0.01718 | 0 |
| 32 | baseline | 56.31/s | — | 19.33 s | 561.4 ms | 659.4 ms | 695.9 ms | 0.12338 | 0.01710 | 0 |
| 32 | torch.compile | 72.22/s | **+28.3% / 1.28×** | 15.07 s | 440.6 ms | 626.7 ms | 755.3 ms | 0.09601 | 0.01740 | 0 |
| 32 | CUDA graphs | 130.12/s | **+131.1% / 2.31×** | 8.36 s | 244.7 ms | 330.2 ms | 394.6 ms | 0.05304 | 0.01718 | 0 |

## Checklist

- [x] Format your code according with pre-commit.
- [x] Add unit tests.
- [x] Update documentation / docstrings / example tutorials as needed.
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed.
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

### CODE REVIEW

File: sglang_omni/models/fun_asr/encoder_cuda_graph.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            static_xs[:b, :t].copy_(xs, non_blocking=True)
+            # Padded rows keep ilens=1: one valid zeroed frame, output dropped.
+            static_ilens.fill_(1)
+            static_ilens[:b].copy_(
+                torch.as_tensor(lengths, dtype=torch.long), non_blocking=True
+            )
+            graph.replay()
+            # Clone before releasing the lock: the next replay overwrites
+            # static_out in place.
+            return static_out[:b].clone()
```
Comment (luojiaxuan) [inline sglang_omni/models/fun_asr/encoder_cuda_graph.py]: The clone is asynchronous, so releasing this CPU lock can let the scheduler stream overwrite static_out before the pre-LM stream finishes copying it; please add cross-stream synchronization or keep the buffer protected until the copy completes.

**Review by luojiaxuan — CHANGES_REQUESTED** [review]
The graph output must remain protected until its copy completes.

File: sglang_omni/models/fun_asr/encoder_cuda_graph.py (hunk 1 above)
Comment (WilsonZheng0327) [reply sglang_omni/models/fun_asr/encoder_cuda_graph.py]: Addressed. Recording a runner-level CUDA event after each output copy and wait on it before the next call touches any static buffer.

File: sglang_omni/models/fun_asr/encoder_cuda_graph.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        self._device = reference.device
+        self._dtype = reference.dtype
+        self._max_batch = max(int(max_batch_size), 1)
+        self._min_free_bytes = int(float(min_free_gb) * (1024**3))
+        self._warmup_iters = int(warmup_iters)
+        # (batch_bucket, t_bucket) -> (graph, static_xs, static_ilens, static_out)
+        self._graphs: dict[Tuple[int, int], tuple] = {}
+        self._failed: set[Tuple[int, int]] = set()
+        self._pool = None
+        # Serializes capture and replay: replay mutates the bucket's static
```
Comment (luojiaxuan) [inline sglang_omni/models/fun_asr/encoder_cuda_graph.py]: add note xxx

File: sglang_omni/models/fun_asr/encoder_cuda_graph.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        )
+        static_ilens = torch.ones(batch_bucket, device=self._device, dtype=torch.long)
+
+        def _masked_forward() -> torch.Tensor:
+            mask = _sanm_mask_from_lengths(
+                static_ilens, t_bucket, dtype=self._dtype, device=self._device
+            )
+            return self._forward(static_xs, mask)
+
+        # Warmup on a fresh stream so allocator state settles before capture.
```
Comment (luojiaxuan) [inline sglang_omni/models/fun_asr/encoder_cuda_graph.py]: ditto

File: sglang_omni/models/fun_asr/encoder_cuda_graph.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        with torch.cuda.stream(stream):
+            for _ in range(self._warmup_iters):
+                _masked_forward()
+        torch.cuda.current_stream().wait_stream(stream)
+        torch.cuda.synchronize()
+
+        if self._pool is None:
+            self._pool = torch.cuda.graph_pool_handle()
+        graph = torch.cuda.CUDAGraph()
+        # thread_local error mode: the LM scheduler thread keeps launching
```
Comment (luojiaxuan) [inline sglang_omni/models/fun_asr/encoder_cuda_graph.py]: ditto

File: sglang_omni/models/fun_asr/encoder_cuda_graph.py (hunk 2 above)
Comment (WilsonZheng0327) [reply sglang_omni/models/fun_asr/encoder_cuda_graph.py]: addressed

File: sglang_omni/models/fun_asr/encoder_cuda_graph.py (hunk 3 above)
Comment (WilsonZheng0327) [reply sglang_omni/models/fun_asr/encoder_cuda_graph.py]: addressed

File: sglang_omni/models/fun_asr/encoder_cuda_graph.py (hunk 4 above)
Comment (WilsonZheng0327) [reply sglang_omni/models/fun_asr/encoder_cuda_graph.py]: addressed

**Review by luojiaxuan — APPROVED** [review]
